tket/src/Gate/Gate.cpp

```diff
@@ -459,6 +459,96 @@ bool Gate::is_clifford() const {
   }
 }
 
+bool Gate::test_exchange_invariance_of_ports(
```

Comment:
This leads to errors in 9 other tests that combine other compilation passes with remove_redundancies(). Some circuits appear to contain gates with n_qubits = 0. Have seen so far with PhasedX and Z gates. Not sure if that is expected behaviour or a bug? Do you know?

Reply:
E.g., in test_SynthesiseTK.cpp, the test starting at line 58, a Z gate with n_qubits = 0 somehow gets created. Appears to happen during other passes because the original circuit does not contain a Z gate.

Reply:
That seems strange, I think we need to understand how that is happening.

Reply:
These n_qubits_=0 gates are coming from the Circuit::dagger() method. I can trace the problem in the test_SynthesiseTK.cpp  back to `Circuit/CircUtils.cpp` line 1249, where this method is called. Also I looked in `tests/Circuit/test_Circ.cpp` and found that the daggered circuit from line 1394 also contains such gates. Not sure how to fix Circuit::dagger() but seems like it might be necessary

Reply:
> These n_qubits_=0 gates are coming from the Circuit::dagger() method. I can trace the problem in the test_SynthesiseTK.cpp back to `Circuit/CircUtils.cpp` line 1249, where this method is called. Also I looked in `tests/Circuit/test_Circ.cpp` and found that the daggered circuit from line 1394 also contains such gates. Not sure how to fix Circuit::dagger() but seems like it might be necessary

This look like a bug unfortunately. I see some places where `get_op_ptr()` is called without specifying the number of qubits, so it defaults to 0. Surprised this hasn't had other consequences, but in any case I think we should fix this.

Reply:
By the way, does `Gate::n_qubits()` (rather than `n_qubits_`) give the expected result?

Reply:
Yes, I think we cannot rely on `n_qubits_`:
```cxx
  unsigned n_qubits_; /**< Number of qubits, when not deducible from type */
```
Should use `n_qubits()` instead.

Reply:
Ok, yeah, `n_qubits()` fixes the problem. Thanks!